flow/layers/physical_shape_layer.cc

```diff
@@ -41,6 +41,10 @@ void PhysicalShapeLayer::Preroll(PrerollContext* context,
     set_paint_bounds(ComputeShadowBounds(path_.getBounds(), elevation_,
                                          context->frame_device_pixel_ratio));
   }
+
+  if (!context->cull_rect.intersects(paint_bounds())) {
+    set_paint_bounds(SkRect::MakeEmpty());
+  }
```

Comment:
I am not sure about this concept. I think it would be better if we keep this simple and report the actual bounds whether or not they intersect the clip. If we want to cull the layers then we should do so outside of this method.

Reply:
Setting empty paint bounds means `needs_painting()` will return false. Other layers that keep empty bounds (if not intersecting clip) also do it in preroll (clip_rect_layer). Where else would you do this?

Reply:
I could modify the code to compute paint bounds and then only set it only if intersecting the clip. So set_paint_bounds would only be called at most once (or not at all if no intersection).

Reply:
The code in clip_rect_layer also gives me pause. If this is part of a retained layer tree that was not clipped before and is now inserted underneath a clip, it will fail to call set_paint_bounds() and the bounds could be anything, so avoiding the call is not the right solution there either. I don't think the behavior will be incorrect since it would just mean extra paint calls.

I've always felt that the right way to handle this concept is to have needs_painting() take the clip bounds as an argument.

Reply:
I like the idea of having extra cull_rect argument in needs_painting instead of setting paint_bounds to empty. I'll look into it.

Reply:
`needs_painting()` taking cull_rect argument is problematic, because we don't track the cull rect during `Paint()`. So we can either store the cull rect during preroll, or set a needs_painting flag during preroll (=paint_bounds.intersects(cull_rect)), or track the cull rect during paint just like during preroll.

But there is one additional thing I realized. ClipRectLayer will not preroll children outside of cull rect. With partial repaint, that would mean those children would not be able to [touch the raster cache](https://github.com/flutter/engine/pull/21498/commits/8e9f6ded04b1b7bd824ce5317539d35bec8db4e0#diff-ec502543ef202d125697cd146e735cffR113). So for example during partial repaint every picture layers in a ClipRectLayer that are outside of currently repainted area would be evicted from cache. I'm not sure if there is way around this other than pre-rolling the whole layer tree and ensuring that every leaf layer is doing the right thing if outside of cull-rect.